### ISSUE TITLE
srvCfg.StmtCfg.Rset.SetNumberBigFloat(ora.F64) does not work

### ISSUE THREAD

**Issue by maliutin** [rana/ora#99]
go version: go1.6.2 windows/amd64
ora commit: 157b7bb2f6ff142e3885412eaaf5eb4a7993b9e7

We have AVAILABLE_BALANCE column of ACCOUNT table. Its type is NUMBER(10, 2)

```SQL
CREATE TABLE ACCOUNT(
...
  AVAILABLE_BALANCE NUMBER(10, 2) NOT NULL,
...
)
```

Now I need that ora provider returns values of AVAILABLE_BALANCE as float64.

I used following code:

```GO
package main

import (
	"fmt"

	ora "gopkg.in/rana/ora.v3"
)

func main() {
	var err error
	var env *ora.Env

	env, err = ora.OpenEnv(nil)
        if err != nil { panic(err) }
	defer env.Close()

	var srvCfg = ora.NewSrvCfg()
	srvCfg.Dblink = "host.com/ORCL"

	// srvCfg.StmtCfg.Rset.SetNumberBigFloat(ora.F64) // does not work.

	var srv *ora.Srv
	srv, err = env.OpenSrv(srvCfg)
	if err != nil { panic(err) }
	defer srv.Close()

       var sesCfg *ora.SesCfg = ora.NewSesCfg()
       sesCfg.Username = "branch"
       sesCfg.Password = "*****"

	var ses *ora.Ses
	ses, err = srv.OpenSes(sesCfg)
	if err != nil { panic(err) }
	defer ses.Close()

	var stmt *ora.Stmt
	stmt, err = ses.Prep("CALL pkg_account.get_account(:1, :2)")
	if err != nil { panic(err) }

	stmt.Cfg().Rset.SetNumberBigFloat(ora.F64)

	rset := &ora.Rset{}
	accountId := int64(100005222)

	_, err = stmt.Exe(rset, accountId)
	if err != nil { panic(err) }

	columnNames := rset.ColumnNames

	row := rset.NextRow()

	for i, name := range columnNames  {
		fmt.Printf("%d, %q, %T, %v\n", i, name, row[i], row[i])
	}

	fmt.Println("successful")
}
```

I wanted to set this option for all request but following does not do it
```GO
srvCfg.StmtCfg.Rset.SetNumberBigFloat(ora.F64)
```
Output:
```
...
10, "AVAILABLE_BALANCE", ora.OCINum, -326300
...
successful
```

If we set the option for the specified statement then we receive value as float64: 
```GO
stmt.Cfg().Rset.SetNumberBigFloat(ora.F64)
```
Output:
```
...
10, "AVAILABLE_BALANCE", float64, -326300
...
successful
```

**Comment by tgulacsi** [comment]
TL;DR; OpenEnv does not use ora.Cfg().

As Far As I See,
1. rset uses rset.stmt.cfg,
2. which is from ses.cfg.StmtCfg,
3. which is from srv.OpenSes(sesCfg),
4. but if sesCfg.StmtCfg == nil, then (srv.go L252) ses.srv.cfg.StmtCfg is
used;
5. srv.cfg is from env.OpenSrv(srvCfg),
6. but if srvCfg.StmtCfg == nil, then (env.go L167) srv.env.cfg.StmtCfg is
used;
7. env.cfg is from ora.OpenEnv(envCfg).

Possible solutions:
a) put the same logic (if envCfg.StmtCfg == nil, then use
_srv.sqlPkgEnv.cfg.StmtCfg) with envCfg,
b) export _drv.sqlPkgEnv as ora.GetDrvEnv()
c) both.

What's your opinion?

Leonid Maliutin <notifications@github.com> ezt írta (időpont: 2016. jún.
23., Cs, 15:12):

**Comment by maliutin** [comment]
Hi, Tamás

```go
if ses.cfg.StmtCfg == nil && ses.srv.cfg.StmtCfg != nil {
	ses.cfg.StmtCfg = &(*ses.srv.cfg.StmtCfg) // copy by value so that user may change independently
}
```
`ses.cfg.StmtCfg == nil` is always `false` because `NewSesCfg` function always create default StmtCfg for new SesCfg.
```go
// NewSrvCfg creates a SrvCfg with default values.
func NewSesCfg() *SesCfg {
	c := &SesCfg{}
	c.StmtCfg = NewStmtCfg()
	return c
}
```

I can fix my code in two ways:
1. Directly create SesCfg structure
```go
var sesCfg *ora.SesCfg = &ora.SesCfg{}
sesCfg.Username = "branch"
sesCfg.Password = "*****"
```
In this case sesCfg.StmtCfg is nil and our condition is `true`.
2. Manually set that `sesCfg.StmtCfg = nil` after `NewSesCfg` function:
```go
var sesCfg *ora.SesCfg = ora.NewSesCfg()
sesCfg.Username = "branch"
sesCfg.Password = "*****"
sesCfg.StmtCfg = nil
```

I think that the creating default SesCfg in `NewSesCfg` function is not clear practice.

I offer to remove `NewSrvCfg` and `NewSesCfg` functions. In this case, the programmer will have possibility to think about StmtCfg field.

**Comment by tgulacsi** [comment]
@rana please help, what's your opinion?

a) don't initialize StmtCfg 'co, leave it to the programmer, and document this;
b) initialize StmtCfg 'co from _drv.sqlPkgEnv..., both in NewSesCfg and NewStmtCfg.

a) is a backwards-incompatible change,
b) could be, but it is done today, too, but not everywhere, so this could be more a clarification than a change. Also, this'd help to be consistent: if sy changes the cfg with ora.Cfg(), maybe she assumes that it is used everywhere in ora from there, but this is not the case ATM.

I'd favor b), what do you think @maliutin ?

**Comment by maliutin** [comment]
I think StmtCfg is needed only in Ses and in Stmt because only these objects associated with data types which are obtained in the queries.
As I see in my project, to understand what type merging is used I should have StmtCfg in separated value and set it when new session is opening.

The current cascade behaviors is not clear for me and programmers that are reviewing my code.

**Comment by tgulacsi** [comment]
PTAL "issue99" branch!
I've eliminated NewSesCfg(), and if you provide a *SesCfg with nil StmtCfg, it will inherit from it srv.
See `TestEnv_SrvCfg` and `TestEnv_SesCfg` in `z_environment_test.go`.

**Comment by tgulacsi** [comment]
I've also removed NewSrvCfg.

**Comment by maliutin** [comment]
Hi. I tested with my project. All work fine.

**Comment by tgulacsi** [comment]
Thanks, merged it to master.
